fri/src/prover.rs

```diff
@@ -95,20 +96,57 @@ where
     let mut data = vec![];
 
     while folded.len() > config.blowup() * config.final_poly_len() {
-        let leaves = RowMajorMatrix::new(folded, 2);
-        let (commit, prover_data) = config.mmcs.commit_matrix(leaves);
+        let arity = config.arity();
+        let next_folded_len = folded.len() / arity;
+
+        // First, we collect the polynomial evaluations that will be committed this round.
+        // Those are `folded` and polynomials in `inputs` not consumed yet with number of
+        // evaluations more than `next_folded_len`
+        let mut polys_before_next_round = vec![];
+
+        let mut cur_folded_len = folded.len();
+        while cur_folded_len > next_folded_len {
+            if let Some(poly_eval) = inputs_iter.next_if(|v| v.len() == cur_folded_len) {
+                let width = poly_eval.len() / next_folded_len;
+                let poly_eval_matrix = RowMajorMatrix::new(poly_eval, width);
+                polys_before_next_round.push(poly_eval_matrix);
+            }
+
+            cur_folded_len /= 2;
```

Comment:
this does not take into account that `inputs_iter` can have multiple elements of the same `v.len()`.
previously the loop relied on `folded.len()` to update instead of explicit `cur_folded_len /= 2`

Reply:
We assume that `inputs` all have distinct lengths and are given in decreasing order. Previously, `folded.len()` was used but `folded.len()` went down by half every iteration

Reply:
that's what I was pointing out -- I'm not sure that's an assumption that must hold? for example see the assert only says the heights must be >=

Reply:
hmm I do see
```
let mut reduced_openings: [Option<Vec<Challenge>>; 32]
let fri_input = reduced_openings.into_iter().rev().flatten().collect_vec();
```
so it does mean the heights strictly decrease, but in fact could decrease by more than 2 at a time (which your calculation does cover I think)